### PR TITLE
core: change msg.content.ptr type to void *

### PR DESCRIPTION
**Pull request by kaspar030**
... makes a lot of casting unnecessary.

### CODE REVIEW

**Comment by miri64** [comment]
As discussed in #5497 ;-)

I'm happy with it. Let's see what Murdock says.

**Comment by miri64** [comment]
Can you maybe provide your sed commands?

**Comment by kaspar030** [comment]
> Can you maybe provide your sed commands?

manual edited these...

**Comment by miri64** [comment]
Some very specific casting errors were found by Murdock.

**Comment by kaspar030** [comment]
> Some very specific casting errors were found by Murdock.

@miri64 could you please double-check the change in zep (the one with the ringbuffer_add)?

**Comment by miri64** [comment]
@kaspar030, no that actually needs to be a pointer of a pointer, since https://github.com/RIOT-OS/RIOT/blob/master/sys/net/gnrc/application_layer/zep/gnrc_zep.c#L814-L815 does the same (correctness of that code not withstanding, might be that there is no sense in it at all, but it used to work so ¯\_(ツ)_/¯)

**Comment by kaspar030** [comment]
@miri64 like this?

**Comment by miri64** [comment]
Except the missing space in the cast, yes. But I'm tired of this discussion so I'm not holding that as a reason not to merge or ACK ;-).

**Comment by miri64** [comment]
Murdock is happy, so ACK. This is a core change (and major API change) so it needs a second ACK.

**Comment by kaspar030** [comment]
> and major API change

While I see the need for a second look, I consider the change minor. All code but three ```printf()``` calls compiled without changing, removal of the actual casts was cosmetic.

File: sys/net/gnrc/application_layer/zep/gnrc_zep.c [hunk 1]
```diff
@@ -573,7 +573,7 @@ void *_event_loop(void *args)
         switch (msg.type) {
             case GNRC_NETAPI_MSG_TYPE_RCV:
                 DEBUG("zep: GNRC_NETAPI_MSG_TYPE_RCV\n");
-                ringbuffer_add(&_rx_buf, (char *)(&msg.content.ptr),
+                ringbuffer_add(&_rx_buf, (void*)&msg.content.ptr,
```
Comment (cgundogan) [inline sys/net/gnrc/application_layer/zep/gnrc_zep.c]: isn't this cast superfluous now?

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c [hunk 2]
```diff
@@ -289,43 +289,43 @@ static void *_event_loop(void *args)
             /* case GNRC_NDP_MSG_ADDR_TIMEOUT: */
             /*     DEBUG("ipv6: Router advertisement timer event received\n"); */
             /*     gnrc_ipv6_netif_remove_addr(KERNEL_PID_UNDEF, */
-            /*                                 (ipv6_addr_t *)msg.content.ptr); */
+            /*                                 msg.content.ptr); */
```
Comment (cgundogan) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: I know it's just a comment, but isn't the former cast necessary in this case?

File: sys/net/gnrc/application_layer/zep/gnrc_zep.c (hunk 1 above)
Comment (kaspar030) [inline sys/net/gnrc/application_layer/zep/gnrc_zep.c]: No, as ```&msg.content.ptr``` has type ```void**```, it needs to be cast to ```void*```.

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c (hunk 2 above)
Comment (kaspar030) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: Yes it is, will fix.

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c [hunk 3]
```diff
@@ -338,18 +338,17 @@ static void *_event_loop(void *args)
 #ifdef MODULE_GNRC_SIXLOWPAN_ND_ROUTER
             case GNRC_SIXLOWPAN_ND_MSG_ABR_TIMEOUT:
                 DEBUG("ipv6: border router timeout event received\n");
-                gnrc_sixlowpan_nd_router_abr_remove(
-                        (gnrc_sixlowpan_nd_router_abr_t *)msg.content.ptr);
+                gnrc_sixlowpan_nd_router_abr_remove(msg.content.ptr);
```
Comment (cgundogan) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: here the cast also seems necessary?

Comment (cgundogan) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: I am not sure actually, will the compiled assume the type of the signature in such instances?

Comment (kaspar030) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: yup

Comment (cgundogan) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: yup to what? (:

File: sys/net/gnrc/application_layer/zep/gnrc_zep.c (hunk 1 above)
Comment (miri64) [inline sys/net/gnrc/application_layer/zep/gnrc_zep.c]: But `ringbuffer_add()` expects `char *`, right?

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c (hunk 3 above)
Comment (kaspar030) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: before: yup to what you said before. but actually, void pointers can be used anywhere. so no changes needed. ;)

Comment (miri64) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: @cgundogan `void *` never needs to be cast, not from nor to. ;-)

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c (hunk 2 above)
Comment (miri64) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: Why is it necessary?

Comment (cgundogan) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: It's not, following your reasoning that casting isn't a MUST

Comment (kaspar030) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: yeah, sorry, got confused.

**Comment by cgundogan** [comment]
Apart from @miri64's comment, ACK

File: sys/net/gnrc/application_layer/zep/gnrc_zep.c (hunk 1 above)
Comment (cgundogan) [inline sys/net/gnrc/application_layer/zep/gnrc_zep.c]: isn't the same true for `void**` as for `void*`? Do you actually need to cast if `void*` would not need it?

**Comment by kaspar030** [comment]
Before we remove all the casts - C++ compilers need them...

**Comment by miri64** [comment]
I though `*.c` files are not compiled by C++ compilers in our current build-system configuration...

**Comment by kaspar030** [comment]
That's correct, and a quick check shows that not many files actually would compile with a c++ compiler, so I think it's ok to drop the casts.

**Comment by miri64** [comment]
Then let's merge this one. (I don't mind the `(void *)` cast in `ringbuffer_add()` since the type of that parameter should be `void *` too IMHO)

**Pull request merged**